File: scripts/festival_updater.py
#!/usr/bin/env python3
"""
Festival Information Updater

This script automatically updates existing festival information.
It uses OpenAI GPT API to gather information from various sources and creates structured updates.

Usage:
    python festival_updater.py --update-existing=true
"""

import argparse
import json
import logging
import os
import sys
from datetime import datetime
from typing import Dict, List

from openai import OpenAI
from dotenv import load_dotenv

# Constants
FESTIVALS_FILE = os.path.join(os.path.dirname(os.path.dirname(__file__)), "db.json")
PROMPT_FILE = os.path.join(os.path.dirname(__file__), "prompt.md")
# OPENAI_MODEL = "gpt-5"
OPENAI_MODEL = "gpt-4.1-mini"
SYSTEM_INSTRUCTIONS = """
You are a data extraction assistant.
Always return valid compact JSON: {"bands":["..."],"ticketPrice":"..."}.
Never include explanations or commentary.
"""


# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)


class FestivalUpdater:
    """Main class for updating festival information."""

    def __init__(self):
        """Initialize the festival updater with necessary clients."""
        self.openai_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
        self.update_existing_festivals_prompt = self._load_prompt_file(PROMPT_FILE)

        # Load current festival data
        self.current_festivals = self._load_festivals()
        self.updates_made = []
        self.new_festivals = []
        self.sources = {}

    def _ask_openai(self, prompt):
        response = self.openai_client.responses.create(
            model=OPENAI_MODEL,
            instructions=SYSTEM_INSTRUCTIONS,
            input=prompt,
            tools=[{"type": "web_search"}],
            temperature=0.3,
        )
        print(f" == Request usage: {response.usage}")
        return response.output_text

    def _load_prompt_file(self, filepath: str) -> str:
        """Load prompt template from a file."""
        try:
            with open(filepath, "r", encoding="utf-8") as f:
                return f.read()
        except FileNotFoundError:
            logger.error(f"Prompt file {filepath} not found")
            return ""
        except Exception as e:
            logger.error(f"Error loading prompt file {filepath}: {e}")
            return ""

    def _load_festivals(self) -> List[Dict]:
        """Load current festival data from db.json."""
        try:
            with open("db.json", "r", encoding="utf-8") as f:
                return json.load(f)["festivals"]
        except FileNotFoundError:
            logger.error("db.json not found")
            return []
        except json.JSONDecodeError as e:
            logger.error(f"Error parsing db.json: {e}")
            return []

    def _save_festivals(self, festivals: List[Dict]) -> None:
        """Save updated festival data to db.json."""
        db_json = { "festivals": festivals }
        with open("db.json", "w", encoding="utf-8") as f:
            json.dump(db_json, f, indent=2, ensure_ascii=False)

    def _search_festival_info(self, festival_name: str, festival_url: str, location: str = "") -> Dict:
        """Search for festival information using OpenAI."""
        prompt_template_filled = (
            self.update_existing_festivals_prompt.replace("{{ FESTIVAL_NAME }}", festival_name)
            .replace("{{ FESTIVAL_LOCATION }}", location)
            .replace("{{ FESTIVAL_URL }}", festival_url)
        )

        try:
            content = self._ask_openai(prompt_template_filled)

            return json.loads(content)

        except Exception as e:
            logger.error(f"Error searching for {festival_name}: {e}")
            return {}

    def update_existing_festivals(self) -> None:
        """Update information for existing festivals."""
        logger.info("Updating existing festivals...")

        for i, festival in enumerate(self.current_festivals):
            festival_name = festival.get("name", "")
            url = festival.get("website", "")
            location = festival.get("location", "")

            logger.info(f"Updating {festival_name}...")

            # Search for updated information
            updated_info = self._search_festival_info(festival_name, url, location)

            if not updated_info:
                continue

            changes = []

            # Update bands if new ones are found
            if updated_info.get("bands") and len(updated_info["bands"]) > len(festival.get("bands", [])):
                old_bands = set(festival.get("bands", []))
                new_bands = set(updated_info["bands"])
                added_bands = new_bands - old_bands

                if added_bands:
                    festival["bands"] = updated_info["bands"]
                    changes.append(f"Added bands: {', '.join(added_bands)}")

            # Update ticket price if available and different
            if updated_info.get("ticketPrice") != "" and updated_info["ticketPrice"] != festival.get("ticketPrice"):
                old_price = festival.get("ticketPrice", "Unknown")
                festival["ticketPrice"] = updated_info["ticketPrice"]
                changes.append(f"Updated ticket price: {old_price}€ → {updated_info['ticketPrice']}€")

            if changes:
                self.updates_made.append({
                    "festival": festival_name,
                    "changes": changes,
                    "sources": updated_info.get("sources", []),
                    "confidence": updated_info.get("confidence", "medium")
                })
                self.current_festivals[i] = festival

    def generate_pr_summary(self) -> str:
        """Generate a summary for the pull request."""
        summary_parts = []

        # Header
        summary_parts.append("# 🎸 Automated Festival Information Update")
        summary_parts.append("")
        summary_parts.append(f"**Update Date:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S UTC')}")
        summary_parts.append("")

        # Summary
        total_changes = len(self.updates_made) + len(self.new_festivals)
        summary_parts.append("## 📊 Summary")
        summary_parts.append(f"- **Existing festivals updated:** {len(self.updates_made)}")
        summary_parts.append(f"- **New festivals added:** {len(self.new_festivals)}")
        summary_parts.append(f"- **Total changes:** {total_changes}")
        summary_parts.append("")

        # Existing festivals updates
        if self.updates_made:
            summary_parts.append("## 🔄 Updated Festivals")
            summary_parts.append("")

            for update in self.updates_made:
                summary_parts.append(f"### {update['festival']}")
                summary_parts.append(f"**Confidence:** {update['confidence']}")
                summary_parts.append("**Changes:**")
                for change in update["changes"]:
                    summary_parts.append(f"- {change}")

                if update["sources"]:
                    summary_parts.append("**Sources:**")
                    for source in update["sources"]:
                        summary_parts.append(f"- {source}")

                summary_parts.append("")

        # New festivals
        if self.new_festivals:
            summary_parts.append("## ✨ New Festivals Added")
            summary_parts.append("")

            for new_festival in self.new_festivals:
                festival = new_festival["festival"]
                summary_parts.append(f"### {festival['name']}")
                summary_parts.append(f"**Location:** {festival['location']}")
                summary_parts.append(f"**Dates:** {festival['dates'].get('start', 'TBD')} to {festival['dates'].get('end', 'TBD')}")
                summary_parts.append(f"**Ticket Price:** €{festival.get('ticketPrice', 'TBD')}")
                summary_parts.append(f"**Website:** {festival.get('website', 'TBD')}")
                summary_parts.append(f"**Confidence:** {new_festival['confidence']}")

                if festival.get("bands"):
                    summary_parts.append(f"**Bands:** {', '.join(festival['bands'][:5])}")
                    if len(festival["bands"]) > 5:
                        summary_parts.append(f" (and {len(festival['bands']) - 5} more)")

                if new_festival["sources"]:
                    summary_parts.append("**Sources:**")
                    for source in new_festival["sources"]:
                        summary_parts.append(f"- {source}")

                summary_parts.append("")

        # Footer
        summary_parts.append("---")
        summary_parts.append("🤖 This update was automatically generated using AI-powered festival information gathering.")
        summary_parts.append("Please review all changes carefully before merging.")

        return "\n".join(summary_parts)

    def run(self, update_existing: bool = True) -> None:
        """Run the festival updater."""
        logger.info("Starting festival information update...")

        if update_existing:
            self.update_existing_festivals()

        if self.updates_made:
            self._save_festivals(self.current_festivals)

            # Generate PR summary
            summary = self.generate_pr_summary()
            with open("festival_update_summary.md", "w", encoding="utf-8") as f:
                f.write(summary)

            logger.info(f"Update complete! {len(self.updates_made)} festivals updated.")
        else:
            logger.info("No updates found.")


def main():
    """Main function."""
    parser = argparse.ArgumentParser(description="Update festival information")
    parser.add_argument("--update-existing", type=str, default="true", help="Update existing festivals")

    args = parser.parse_args()

    # Convert string boolean arguments
    update_existing = args.update_existing.lower() == "true"

    # Load all environment variables from .env file if exists
    load_dotenv()

    # Check for required environment variables
    if not os.getenv("OPENAI_API_KEY"):
        logger.error("OPENAI_API_KEY environment variable is required")
        sys.exit(1)

    try:
        updater = FestivalUpdater()
        updater.run(update_existing=update_existing)
    except Exception as e:
        logger.error(f"Error running festival updater: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/festival_updater/festival_updater.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/neovasili/metal-fests/internal/data"
	"github.com/neovasili/metal-fests/internal/model"
	"github.com/neovasili/metal-fests/internal/openai"
)

type FestivalUpdateResult struct {
	Bands       []model.BandRef `json:"bands"`
	TicketPrice *float64        `json:"ticketPrice"`
}

type FestivalChange struct {
	Name         string
	NewBands     []string
	UpdatedBands []string
	OldPrice     float64
	NewPrice     float64
	PriceUpdated bool
}

type UpdateStats struct {
	TotalFestivals   int
	UpdatedFestivals int
	NewBands         int
	UpdatedPrices    int
	TotalTokens      int
	TotalCost        float64
	UsedModel        string
	PromptTokens     int
	CompletionTokens int
	Changes          []FestivalChange
}

var openaiClient *openai.OpenAIClient

func searchFestivalInfo(promptTemplate string, festival model.Festival, useFallbackModel bool, dryRun bool) (*FestivalUpdateResult, int, float64, string, error) {
	userPrompt := strings.ReplaceAll(promptTemplate, "{{ FESTIVAL_NAME }}", festival.Name)
	userPrompt = strings.ReplaceAll(userPrompt, "{{ FESTIVAL_LOCATION }}", festival.Location)
	userPrompt = strings.ReplaceAll(userPrompt, "{{ FESTIVAL_URL }}", festival.Website)

	usedTokens := 0
	estimatedCost := 0.0
	usedModel := ""

	var festivalJsonSchema = map[string]any{
		"type":                 "object",
		"additionalProperties": false,
		"properties": map[string]any{
			"bands": map[string]any{
				"type":                 "array",
				"additionalProperties": false,
				"items": map[string]any{
					"type":                 "object",
					"additionalProperties": false,
					"properties": map[string]any{
						"name": map[string]any{"type": "string", "additionalProperties": false},
						"size": map[string]any{"type": "integer", "additionalProperties": false},
					},
					"required": []string{"name", "size"},
				},
			},
			"ticketPrice": map[string]any{
				"type":                 []any{"integer", "null"},
				"additionalProperties": false,
			},
		},
		"required": []string{"bands", "ticketPrice"},
	}

	modelToUse := openai.PrimaryModel
	if useFallbackModel {
		modelToUse = openai.FallbackModel
	}

	resp, err := openaiClient.AskOpenAI(userPrompt, festivalJsonSchema, modelToUse, dryRun)
	if err != nil {
		return nil, usedTokens, estimatedCost, usedModel, err
	}

	if resp == nil {
		return nil, usedTokens, estimatedCost, usedModel, fmt.Errorf("no response from OpenAI")
	}
	usedTokens = resp.TotalUsedTokens
	estimatedCost = resp.EstimatedCost
	usedModel = string(resp.UsedModel)
	fmt.Printf("🧠 Used model: %s\n", usedModel)
	fmt.Printf("📊 Tokens used: %d\n", usedTokens)
	fmt.Printf("💰 Estimated cost: $%.2f\n", estimatedCost)
	if len(resp.OutputText) == 0 {
		return nil, usedTokens, estimatedCost, usedModel, fmt.Errorf("empty response from OpenAI")
	}

	content := strings.TrimSpace(resp.OutputText)
	var result FestivalUpdateResult
	if err := json.Unmarshal([]byte(content), &result); err != nil {
		return nil, usedTokens, estimatedCost, usedModel, fmt.Errorf("failed to parse OpenAI response: %w", err)
	}

	return &result, usedTokens, estimatedCost, usedModel, nil
}

func updateExistingFestivals(promptTemplate string, dryRun bool, festivalName string, openaiResponseFilePath string) *UpdateStats {
	festivals, err := data.GetFestivals()
	if err != nil {
		fmt.Printf("  ⚠️  Error fetching festivals: %v\n", err)
	}

	// Filter by festival name if provided
	if festivalName != "" {
		filteredFestivals := make([]model.Festival, 0)
		for _, festival := range festivals {
			if festival.Name == festivalName {
				filteredFestivals = append(filteredFestivals, festival)
				break
			}
		}
		if len(filteredFestivals) == 0 {
			fmt.Printf("  ⚠️  Festival '%s' not found\n", festivalName)
			return &UpdateStats{}
		}
		festivals = filteredFestivals
	}

	stats := &UpdateStats{
		TotalFestivals: len(festivals),
	}

	fmt.Printf("Updating %d festivals...\n", stats.TotalFestivals)

	for index, festival := range festivals {
		fmt.Printf("\n[%d/%d] Processing %s...\n", index+1, stats.TotalFestivals, festival.Name)

		var result = &FestivalUpdateResult{}
		var tokens int
		var cost float64
		var usedModel string
		if openaiResponseFilePath != "" && festivalName != "" {
			// Load OpenAI response from file for testing
			// #nosec G304 -- This is a command-line script where the file path is provided by the user
			content, err := os.ReadFile(openaiResponseFilePath)
			if err != nil {
				fmt.Printf("  ⚠️  Error reading OpenAI response file: %v\n", err)
				continue
			}
			fmt.Println(string(content))
			if err := json.Unmarshal(content, result); err != nil {
				fmt.Printf("  ⚠️  Error parsing OpenAI response file: %v\n", err)
				continue
			}
			fmt.Printf("🧠 Loaded OpenAI response from file: %s\n", openaiResponseFilePath)
		} else {
			result, tokens, cost, usedModel, err = searchFestivalInfo(promptTemplate, festival, false, dryRun)
			stats.TotalTokens += tokens
			stats.TotalCost += cost
			stats.UsedModel = usedModel
			if err != nil {
				fmt.Printf("  ⚠️  Error: %v\n", err)
				continue
			}
			if result == nil {
				fmt.Println("  ℹ️  Dry-run mode: skipping update")
				continue
			}
			// If no bands or ticket price found, retry with fallback model
			if len(result.Bands) == 0 && result.TicketPrice == nil {
				result, tokens, cost, usedModel, err = searchFestivalInfo(promptTemplate, festival, true, dryRun)
				stats.TotalTokens += tokens
				stats.TotalCost += cost
				stats.UsedModel = usedModel
				if err != nil {
					fmt.Printf("  ⚠️  Error: %v\n", err)
					continue
				}
			}
		}

		updated := false
		festivalChange := FestivalChange{
			Name: festival.Name,
		}

		// Update bands if new ones found
		if len(result.Bands) > 0 {
			oldBandCount := len(festival.Bands)
			newBands := make([]model.BandRef, 0)
			updatedBands := make([]model.BandRef, 0)

			for _, bandRef := range result.Bands {
				normalizedBandName := data.NormalizeBandName(bandRef.Name)
				band := model.BandRef{
					Key:  data.GenerateBandKey(normalizedBandName),
					Name: normalizedBandName,
					Size: bandRef.Size,
				}
				if !containsBand(festival.Bands, band.Name) {
					newBands = append(newBands, model.BandRef{Key: band.Key, Name: band.Name, Size: band.Size})
					festivalChange.NewBands = append(festivalChange.NewBands, band.Name)
				} else {
					if bandHasBeenUpdated(festival.Bands, band) {
						updatedBands = append(updatedBands, band)
						festivalChange.UpdatedBands = append(festivalChange.UpdatedBands, band.Name)
					}
				}
			}

			if len(newBands) > 0 {
				festival.Bands = append(festival.Bands, newBands...)
				stats.NewBands += len(newBands)
				updated = true
				fmt.Printf("  ✓ Added %d new bands (total: %d → %d)\n", len(newBands), oldBandCount, len(festival.Bands))
			} else {
				fmt.Printf("  - No new bands to add\n")
			}

			if len(updatedBands) > 0 {
				for _, bandRef := range updatedBands {
					festival = updateBandData(festival, bandRef)
				}
				fmt.Printf("  ✓ Updated %d existing bands\n", len(updatedBands))
				updated = true
			}
		}

		// Update ticket price if available and different
		if result.TicketPrice != nil && festival.TicketPrice == 0 {
			festivalChange.OldPrice = festival.TicketPrice
			festivalChange.NewPrice = *result.TicketPrice
			festivalChange.PriceUpdated = true
			oldPrice := fmt.Sprintf("%.2f€", festival.TicketPrice)
			festival.TicketPrice = *result.TicketPrice
			stats.UpdatedPrices++
			updated = true
			fmt.Printf("  ✓ Updated ticket price: %s → %.2f€\n", oldPrice, *result.TicketPrice)
		}

		if updated {
			stats.UpdatedFestivals++
			stats.Changes = append(stats.Changes, festivalChange)
			err = data.UpdateFestivalInDatabase(festival)
			if err != nil {
				fmt.Printf("  ⚠️  Error updating festival in database: %v\n", err)
			}
		}
	}

	return stats
}

// Check if a band is already in the festival's band list
func containsBand(bands []model.BandRef, bandName string) bool {
	for _, band := range bands {
		if band.Name == bandName {
			return true
		}
	}
	return false
}

// Check if a band has been updated
func bandHasBeenUpdated(bands []model.BandRef, band model.BandRef) bool {
	for _, b := range bands {
		if b.Key == band.Key {
			if b.Name != band.Name || b.Size != band.Size {
				return true
			}
		}
	}
	return false
}

// Updates a band data in a festival's band list
func updateBandData(festival model.Festival, band model.BandRef) model.Festival {
	for i, b := range festival.Bands {
		if b.Key == band.Key {
			festival.Bands[i] = model.BandRef{
				Key:  band.Key,
				Name: band.Name,
				Size: band.Size,
			}
			break
		}
	}
	return festival
}

func generatePRSummary(stats *UpdateStats) string {
	var buf bytes.Buffer

	buf.WriteString("# 🤖 Automated Festival Information Update\n\n")
	buf.WriteString("This PR contains automated updates to festival information.\n\n")
	buf.WriteString("## 📊 Update Statistics\n\n")
	buf.WriteString(fmt.Sprintf("- **Total Festivals Processed**: %d\n", stats.TotalFestivals))
	buf.WriteString(fmt.Sprintf("- **Festivals Updated**: %d\n", stats.UpdatedFestivals))
	buf.WriteString(fmt.Sprintf("- **New Bands Added**: %d\n", stats.NewBands))
	buf.WriteString(fmt.Sprintf("- **Ticket Prices Updated**: %d\n", stats.UpdatedPrices))
	buf.WriteString("\n## 🤖 AI Usage Statistics\n\n")
	buf.WriteString(fmt.Sprintf("- **Total Tokens**: %d\n", stats.TotalTokens))
	buf.WriteString(fmt.Sprintf("- **Total Cost**: %.2f €\n", stats.TotalCost))
	buf.WriteString(fmt.Sprintf("- **Model**: %s\n", stats.UsedModel))
	buf.WriteString("\n## ⚙️ Automation Details\n\n")
	buf.WriteString(fmt.Sprintf("- **Run Date**: %s\n", time.Now().Format("2006-01-02 15:04:05 UTC")))
	buf.WriteString("- **Source**: GitHub Actions Workflow\n")
	buf.WriteString("- **Script**: `scripts/festival_updater.go`\n")

	// Add detailed changes section if there are any updates
	if len(stats.Changes) > 0 {
		buf.WriteString("\n<details>\n<summary>📋 Detailed Festival Changes</summary>\n\n")
		for _, change := range stats.Changes {
			buf.WriteString(fmt.Sprintf("### %s\n\n", change.Name))

			if len(change.NewBands) > 0 {
				buf.WriteString(fmt.Sprintf("- **New Bands Added** (%d):\n", len(change.NewBands)))
				for _, band := range change.NewBands {
					buf.WriteString(fmt.Sprintf("  - %s\n", band))
				}
			}

			if len(change.UpdatedBands) > 0 {
				buf.WriteString(fmt.Sprintf("- **Existing Bands Updated** (%d):\n", len(change.UpdatedBands)))
				for _, band := range change.UpdatedBands {
					buf.WriteString(fmt.Sprintf("  - %s\n", band))
				}
			}

			if change.PriceUpdated {
				buf.WriteString(fmt.Sprintf("- **Ticket Price Updated**: %.2f€ → %.2f€\n", change.OldPrice, change.NewPrice))
			}

			buf.WriteString("\n")
		}
		buf.WriteString("</details>\n")
	}

	if stats.UpdatedFestivals == 0 {
		buf.WriteString("\n---\n")
		buf.WriteString("*No updates were needed. All festival information is up to date.*\n")
	} else {
		buf.WriteString("\n---\n")
		buf.WriteString("*This PR was automatically generated. Please review the changes before merging.*\n")
	}

	return buf.String()
}

func main() {
	// Parse command line flags
	dryRun := false
	festivalName := ""
	openaiResponseFilePath := ""

	flag.BoolVar(&dryRun, "dry-run", false, "Enable dry run mode")
	flag.StringVar(&festivalName, "festival", "", "Specify festival name")
	flag.StringVar(&openaiResponseFilePath, "openai-response", "", "Specify OpenAI response file path for testing")
	flag.Parse()

	if dryRun {
		fmt.Println("🔍 DRY-RUN MODE: No API calls will be made, no files will be modified")
		fmt.Println()
	}

	if festivalName != "" {
		fmt.Printf("🎯 Single festival mode: %s\n\n", festivalName)
	}

	apiKey := os.Getenv("OPENAI_API_KEY")
	if apiKey == "" && !dryRun {
		fmt.Fprintf(os.Stderr, "Error: OPENAI_API_KEY environment variable not set\n")
		os.Exit(1)
	}

	openaiClient = openai.NewOpenAIClient(apiKey)

	// Load prompt template
	promptTemplate, err := openai.LoadPromptFile("scripts/festival_prompt.md")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading prompt template: %v\n", err)
		os.Exit(1)
	}

	// Update festivals
	stats := updateExistingFestivals(promptTemplate, dryRun, festivalName, openaiResponseFilePath)

	// Generate PR summary
	summary := generatePRSummary(stats)
	if err := os.WriteFile("festival_update_summary.md", []byte(summary), 0600); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing summary: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("\n✅ Festival update completed successfully!")
	fmt.Printf("📄 Summary written to festival_update_summary.md\n")
}
